win32ss/gdi/ntgdi/freetype.c

```diff
@@ -4731,22 +4778,33 @@ TextIntRealizeFont(HFONT FontHandle, PTEXTOBJ pTextObj)
     }
     else
     {
-        UNICODE_STRING FaceName;
+        UNICODE_STRING Name;
         PFONTGDI FontGdi = ObjToGDI(TextObj->Font, FONT);
+        PSHARED_FACE SharedFace = FontGdi->SharedFace;
 
         IntLockFreeType();
         IntRequestFontSize(NULL, FontGdi, pLogFont->lfWidth, pLogFont->lfHeight);
         IntUnLockFreeType();
 
-        RtlInitUnicodeString(&FaceName, NULL);
-        IntGetFontLocalizedName(&FaceName, FontGdi->SharedFace, TT_NAME_ID_FONT_FAMILY, gusLanguageID);
-
-        /* truncated copy */
-        FaceName.Length = (USHORT)min(FaceName.Length, (LF_FACESIZE - 1) * sizeof(WCHAR));
-        FaceName.MaximumLength = (USHORT)(FaceName.Length + sizeof(UNICODE_NULL));
-        RtlCopyMemory(TextObj->FaceName, FaceName.Buffer, FaceName.MaximumLength);
+        TextObj->TextFace[0] = UNICODE_NULL;
+        if (MatchFontNames(SharedFace, SubstitutedLogFont.lfFaceName))
+        {
+            RtlStringCchCopyW(TextObj->TextFace, _countof(TextObj->TextFace), pLogFont->lfFaceName);
+        }
+        else
+        {
+            RtlInitUnicodeString(&Name, NULL);
+            Status = IntGetFontLocalizedName(&Name, SharedFace, TT_NAME_ID_FONT_FAMILY, gusLanguageID);
+            if (NT_SUCCESS(Status))
+            {
+                /* truncated copy */
+                Name.Length = (USHORT)min(Name.Length, (LF_FACESIZE - 1) * sizeof(WCHAR));
+                RtlCopyMemory(TextObj->TextFace, Name.Buffer, Name.Length);
+                TextObj->TextFace[Name.Length / sizeof(WCHAR)] = UNICODE_NULL;
```

Comment:
No, "Name.Buffer" can be not NULL-terminated. I don't agree.

Reply:
https://docs.microsoft.com/en-us/windows/desktop/api/strsafe/nf-strsafe-stringcbcopynw
```txt
pszSrc
Type: LPCTSTR
The source string. This string must be null-terminated.
```
Name.Buffer can be not null-terminated.

Reply:
Don't fix.

Reply:
I never said that "Name.Buffer" had to be NULL-terminated, I've just said that "TextObj->TextFace" is NULL-terminated.

Reply:
"The source string. This string must be null-terminated." <-- actually, due to the fact it uses the last parameter as the number of bytes to copy, the restriction quoted doesn't really apply (

Reply:
The 3rd parameter of RtlStringCbCopyNW should be null-terminated as the document says.

Reply:
This does not make sense.
What is the difference between copying for example, 3 characters (== 6 bytes when using WCHARs) from a 12321-chars-long NULL-terminated string, and copying 3 characters from a 123-chars-long non-NULL-terminated string ?
The answer is none, because you always stop at the 3rd character, and, in the first case (12321-chars-long NULL-terminated string), how can you differentiate this from a 123-chars-long non-NULL-terminated string followed by garbage ?

Reply:
I see. I obey you. OK.

Reply:
FWIW, I preferred the previous code. It's an obvious pattern that happens in tons of places with UNICODE_STRINGs.
But if you're going to use CopyN, it should use `RTL_NUMBER_OF(DestinationBuffer)` and not some calculated-on-the-fly value that's not guaranteed to be correct.

Reply:
In this case:
`RtlStringCbCopyNW(TextObj->TextFace, sizeof(TextObj->TextFace), Name.Buffer, Name.Length);`
This would allow also to remove the `Name.Length = min()` calculation.
Due to how TextFace is defined in its struct, the sizeof would be equal to LF_FACENAME * sizeof(WCHAR), which is correct.